src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,296 @@
+            /// document (inclusive), an array of single element will be returned for non-linked document.
+            /// </summary>
+            private static bool IsSupportedLinkedDocument(Document document, out ImmutableArray<DocumentId> allDocumentIds)
+            {
+                var solution = document.Project.Solution;
+                var linkedDocumentids = document.GetLinkedDocumentIds();
+
+                // TODO: figure out how to properly determine if and how a document is linked using project system.
+
+                // If we found a linked document which is part of a project with differenct project file,
+                // then it's an actual linked file (i.e. not a multi-targeting project). We don't support that, because 
+                // we don't know which default namespace and folder path we should use to construct target
+                // namespace.
+                if (linkedDocumentids.Any(id => 
+                        !PathUtilities.PathsEqual(solution.GetDocument(id).Project.FilePath, document.Project.FilePath)))
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+
+                // Now determine if the actual file path matches its logical path in project 
+                // which is constructed as <project root path>\Logical\Folders\. The refactoring 
+                // is triggered only when the two match. The reason of doing this is we don't really know
+                // the user's intention of keeping the file path out-of-sync with its logical path.
+                var projectRoot = PathUtilities.GetDirectoryName(document.Project.FilePath);
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                if (PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath))
+                {
+                    allDocumentIds = linkedDocumentids.Add(document.Id);
+                    return true;
+                }
+                else
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+            }
+
+            private static string GetDefaultNamespace(ImmutableArray<Document> documents, ISyntaxFactsService syntaxFacts)
+            {
+                // For all projects containing all the linked documents, bail if 
+                // 1. Any of them doesn't have default namespace, or
+                // 2. Multiple default namespace are found. (this might be possible by tweaking project file).
+                // The refactoring depends on a single default namespace to operate.
+                var defaultNamespaceFromProjects = new HashSet<string>(
```

Comment:
if we are okay with just using hashset, not sure why we are using pooled one above. since it is code tranformation, seems fine not to use any pool since it will only run when user actually invoke it. (as long as fix is calculated lazily as it supposed to)

Reply:
I think the code fix is calculate lazily, but this particular usage is not: it's executed whenever the refactoring is asked for available action. Will change to pooled.